File: coding_part/wro2025_vehicle/src/actuators/vehicle_controller.cpp
// actuators/vehicle_controller.cpp
// Implementation of VehicleController

#include "vehicle_controller.h"
#include <iostream>
#include <cmath>  // For round, clamp functions if needed
#include <thread> // For potential small delays if needed
#include <chrono> // For potential small delays if needed

VehicleController::VehicleController()
    : parkingActive(false), maneuverInProgress(false), parkingState(ParkingState::NOT_PARKING), parkingStepCounter(0)
{
    // Constructor initializes atomic flags and member variables
}

bool VehicleController::initialize(std::shared_ptr<SerialHandler> serialHandler)
{
    if (!serialHandler)
    {
        std::cerr << "VehicleController: Invalid SerialHandler provided." << std::endl;
        return false;
    }
    this->serialHandler = serialHandler;
    std::cout << "VehicleController: Initialized with SerialHandler." << std::endl;
    return true;
}

void VehicleController::executeCommand(const NavigationCommand &command)
{
    // If a complex maneuver like parking is active, prioritize its internal logic
    // unless it's an emergency stop.
    if (isManeuverActive() && command.action != NavigationCommand::Action::EMERGENCY_STOP)
    {
        if (parkingActive)
        {
            updateParkingSequence(); // Continue parking logic
        }
        // Ignore other commands while a maneuver is active
        return;
    }

    // Handle the command based on its action
    switch (command.action)
    {
    case NavigationCommand::Action::FOLLOW_LEFT_LANE:
        handleFollowLeftLane(command.desiredSpeed);
        break;
    case NavigationCommand::Action::FOLLOW_RIGHT_LANE:
        handleFollowRightLane(command.desiredSpeed);
        break;
    case NavigationCommand::Action::GO_STRAIGHT:
        handleGoStraight(command.desiredSpeed);
        break;
    case NavigationCommand::Action::STOP_IN_START:
        handleStopInStart();
        break;
    case NavigationCommand::Action::EXECUTE_PARKING:
        handleExecuteParking(); // This starts the parking sequence
        break;
    case NavigationCommand::Action::EMERGENCY_STOP:
        handleEmergencyStop();
        break;
    default:
        // Unknown action, treat as stop or go straight with zero speed?
        std::cerr << "VehicleController: Unknown NavigationCommand action." << std::endl;
        sendMotorCommand(0.0f);
        sendServoCommand(0.0f); // Center steering
        break;
    }
}

void VehicleController::startParkingSequence()
{
    if (isManeuverActive())
    {
        std::cerr << "VehicleController: Cannot start parking, another maneuver is active." << std::endl;
        return;
    }
    parkingActive.store(true);
    maneuverInProgress.store(true);
    parkingState.store(ParkingState::APPROACHING_SPOT);
    parkingStepCounter = 0;
    std::cout << "VehicleController: Starting parking sequence." << std::endl;
}

void VehicleController::updateParkingSequence()
{
    if (!parkingActive)
        return;

    // Simple state machine for parking
    // In a real implementation, this would be much more sophisticated,
    // using encoders, IMU, or precise timing/distance estimates.
    // This is a basic placeholder sequence.

    // Increment step counter (this would ideally be based on time or distance)
    parkingStepCounter++;

    switch (parkingState.load())
    {
    case ParkingState::APPROACHING_SPOT:
        // 1. Approach the spot slowly, aligning the rear axle with the spot start
        sendMotorCommand(0.3f); // Slow forward
        sendServoCommand(0.0f); // Straight
        if (parkingStepCounter > 50)
        { // Placeholder condition
            parkingState.store(ParkingState::REVERSING_IN);
            parkingStepCounter = 0;
            std::cout << "VehicleController: Parking - Reversing into spot." << std::endl;
        }
        break;

    case ParkingState::REVERSING_IN:
        // 2. Start reversing, turn wheels hard right (or left, depending on side)
        sendMotorCommand(-0.4f); // Slow reverse
        sendServoCommand(1.0f);  // Hard right turn (assuming 1.0 is right)
        if (parkingStepCounter > 100)
        { // Placeholder condition
            parkingState.store(ParkingState::TURNING_IN);
            parkingStepCounter = 0;
            std::cout << "VehicleController: Parking - Turning wheels in." << std::endl;
        }
        break;

    case ParkingState::TURNING_IN:
        // 3. Turn wheels hard left (or right) to align inside the spot
        sendMotorCommand(-0.3f); // Continue slow reverse
        sendServoCommand(-1.0f); // Hard left turn
        if (parkingStepCounter > 50)
        { // Placeholder condition
            parkingState.store(ParkingState::REVERSING_IN_FINAL);
            parkingStepCounter = 0;
            std::cout << "VehicleController: Parking - Final reverse." << std::endl;
        }
        break;

    case ParkingState::REVERSING_IN_FINAL:
        // 4. Continue reversing to pull fully into the spot
        sendMotorCommand(-0.3f); // Slow reverse
        sendServoCommand(-1.0f); // Keep left turn
        if (parkingStepCounter > 80)
        { // Placeholder condition
            parkingState.store(ParkingState::TURNING_STRAIGHT);
            parkingStepCounter = 0;
            std::cout << "VehicleController: Parking - Straightening wheels." << std::endl;
        }
        break;

    case ParkingState::TURNING_STRAIGHT:
        // 5. Straighten the wheels
        sendMotorCommand(0.0f); // Stop
        sendServoCommand(0.0f); // Center steering
        // Small delay to let steering settle?
        if (parkingStepCounter > 20)
        { // Placeholder condition
            parkingState.store(ParkingState::PARKING_COMPLETE);
            parkingStepCounter = 0;
            std::cout << "VehicleController: Parking - Sequence complete." << std::endl;
        }
        break;

    case ParkingState::PARKING_COMPLETE:
        // Parking finished successfully
        parkingActive.store(false);
        maneuverInProgress.store(false);
        parkingStepCounter = 0;
        // Ensure stopped
        sendMotorCommand(0.0f);
        sendServoCommand(0.0f);
        std::cout << "VehicleController: Parking completed successfully." << std::endl;
        break;

    case ParkingState::PARKING_FAILED:
        // Parking failed, stop and exit sequence
        parkingActive.store(false);
        maneuverInProgress.store(false);
        parkingStepCounter = 0;
        sendMotorCommand(0.0f);
        sendServoCommand(0.0f);
        std::cout << "VehicleController: Parking failed." << std::endl;
        break;

    default:
        // Should not happen
        parkingActive.store(false);
        maneuverInProgress.store(false);
        sendMotorCommand(0.0f);
        sendServoCommand(0.0f);
        std::cerr << "VehicleController: Unexpected parking state." << std::endl;
        break;
    }
}

bool VehicleController::isManeuverActive() const
{
    return maneuverInProgress.load();
}

void VehicleController::emergencyStop()
{
    sendMotorCommand(0.0f);
    sendServoCommand(0.0f);
    // Cancel any active maneuvers
    parkingActive.store(false);
    maneuverInProgress.store(false);
    parkingState.store(ParkingState::NOT_PARKING);
    parkingStepCounter = 0;
    std::cout << "VehicleController: Emergency stop executed." << std::endl;
}

// --- Private Helper Functions ---

void VehicleController::sendMotorCommand(float normalizedSpeed)
{
    // Clamp speed to valid range
    if (normalizedSpeed > 1.0f)
        normalizedSpeed = 1.0f;
    if (normalizedSpeed < -1.0f)
        normalizedSpeed = -1.0f;

    // Map normalized speed (-1.0 to 1.0) to Nano's motor command range (-128 to 127)
    int8_t nanoSpeed = static_cast<int8_t>(std::round(normalizedSpeed * 127.0f));
    if (serialHandler)
    {
        serialHandler->sendMotorSpeedCommand(nanoSpeed);
    }
}

void VehicleController::sendServoCommand(float normalizedSteering)
{
    // Clamp steering to valid range
    if (normalizedSteering > 1.0f)
        normalizedSteering = 1.0f;
    if (normalizedSteering < -1.0f)
        normalizedSteering = -1.0f;

    // Map normalized steering (-1.0 to 1.0) to servo angle (0 to 180)
    // Assuming -1.0 = 45 degrees (hard left), 0.0 = 90 degrees (center), 1.0 = 135 degrees (hard right)
    // This mapping can be adjusted based on vehicle characteristics
    uint8_t servoAngle = static_cast<uint8_t>(std::round(90.0f + normalizedSteering * 45.0f));
    // Ensure angle is within 0-180
    if (servoAngle > 180)
        servoAngle = 180;
    // if (servoAngle < 0)
    //     servoAngle = 0;

    if (serialHandler)
    {
        serialHandler->sendServoAngleCommand(servoAngle);
    }
}

// --- Action Handlers ---

void VehicleController::handleFollowLeftLane(float desiredSpeed)
{
    // This is a simplified interpretation.
    // A more advanced system might use a PID controller based on sensor/edge data.
    // For now, assume a basic steering command to follow the left.
    sendMotorCommand(desiredSpeed);
    sendServoCommand(-0.3f); // Slight left turn command
}

void VehicleController::handleFollowRightLane(float desiredSpeed)
{
    sendMotorCommand(desiredSpeed);
    sendServoCommand(0.3f); // Slight right turn command
}

void VehicleController::handleGoStraight(float desiredSpeed)
{
    sendMotorCommand(desiredSpeed);
    sendServoCommand(0.0f); // Center steering
}

void VehicleController::handleStopInStart()
{
    // Command to stop the vehicle
    sendMotorCommand(0.0f);
    sendServoCommand(0.0f);
    // The logic for *when* to call this and ensuring it stops in the correct place
    // should primarily reside in main.cpp using PositionalMemory.
    std::cout << "VehicleController: Executing STOP command." << std::endl;
}

void VehicleController::handleExecuteParking()
{
    // This command initiates the parking sequence
    startParkingSequence();
}

void VehicleController::handleEmergencyStop()
{
    emergencyStop();
}

File: coding_part/wro2025_vehicle_python/actuators/vehicle_controller.py
# actuators/vehicle_controller.py
"""Controls the actuators (motors, servos) via commands sent to the Nano."""

import logging
import time
from enum import IntEnum
from typing import TYPE_CHECKING, Optional
from control.navigation_command import NavigationAction
if TYPE_CHECKING:
    from communication.serial_handler import SerialHandler
    from control.navigation_command import NavigationCommand, NavigationAction
    from sensors.sensor_data import SensorData

logger = logging.getLogger(__name__)

class ParkingState(IntEnum):
    """Represents the state within the parking sequence."""
    APPROACHING_SPOT = 0
    REVERSING_INTO_SPOT = 1
    ADJUSTING_POSITION = 2
    PARKING_COMPLETE = 3
    PARKING_FAILED = 4

class VehicleController:
    """
    Translates high-level NavigationCommands into low-level actuator commands
    sent via the SerialHandler. Manages complex maneuvers like parking.
    """

    def __init__(self, serial_handler: 'SerialHandler'):
        if not serial_handler:
            raise ValueError("VehicleController requires a valid SerialHandler.")
        self.serial_handler = serial_handler

        # Parking state management
        self.parking_active = False
        self.parking_state: ParkingState = ParkingState.APPROACHING_SPOT
        self.parking_step_counter = 0
        self.parking_start_time = 0.0

        # Last sent commands for potential logging or smoothing
        self._last_motor_speed = 0.0
        self._last_servo_angle = 90 # Assume 90 is center

        logger.info("VehicleController initialized.")

    def execute_command(self, command: 'NavigationCommand'):
        """
        Execute a NavigationCommand by sending appropriate motor/servo commands.
        This is the primary interface for the main loop.
        """
        if not command:
            logger.warning("VehicleController: Received None command. Ignoring.")
            return

        action = command.action
        desired_speed = command.desired_speed

        # Handle complex maneuvers first
        if self.parking_active:
            self._execute_parking_step()
            return # Parking takes precedence

        # Map NavigationAction to motor/servo commands
        if action == NavigationAction.GO_STRAIGHT:
            # Speed determines magnitude, servo centered
            motor_speed = self._map_normalized_speed(desired_speed)
            servo_angle = 90 # Straight
            self._send_motor_command(motor_speed)
            self._send_servo_command(servo_angle)

        elif action in (NavigationAction.TURN_LEFT, NavigationAction.FOLLOW_LEFT_LANE):
            motor_speed = self._map_normalized_speed(desired_speed)
            # Negative angle for left turn (assuming 0 is full left, 180 is full right)
            servo_angle = max(0, 90 - 30) # Adjust turn angle as needed
            self._send_motor_command(motor_speed)
            self._send_servo_command(servo_angle)

        elif action in (NavigationAction.TURN_RIGHT, NavigationAction.FOLLOW_RIGHT_LANE):
            motor_speed = self._map_normalized_speed(desired_speed)
            # Positive angle for right turn
            servo_angle = min(180, 90 + 30) # Adjust turn angle as needed
            self._send_motor_command(motor_speed)
            self._send_servo_command(servo_angle)

        elif action == NavigationAction.STOP_IN_START:
            # Gradual stop or immediate stop?
            self._send_motor_command(0)
            self._send_servo_command(90) # Center steering
            logger.info("VehicleController: Executing STOP_IN_START.")

        elif action == NavigationAction.EXECUTE_PARKING:
            # Initiate the parking sequence
            self.start_parking_sequence()

        elif action == NavigationAction.EMERGENCY_STOP:
            self.emergency_stop()
        else:
            logger.warning(f"VehicleController: Unknown action {action}. Stopping.")
            self._send_motor_command(0)
            self._send_servo_command(90)

    def _map_normalized_speed(self, normalized_speed: float) -> int:
        """
        Map a normalized speed (0.0 to 1.0) to an integer motor command (-255 to 255).
        Negative values indicate reverse.
        """
        # Clamp input
        clamped_speed = max(-1.0, min(1.0, normalized_speed))
        # Map to motor range (assuming -255 to 255)
        mapped_speed = int(clamped_speed * 255)
        # Clamp output to motor range if library expects it (pyserial doesn't care)
        return max(-255, min(255, mapped_speed))

    def _send_motor_command(self, speed: int):
        """Send motor speed command via serial handler."""
        if speed != self._last_motor_speed:
            logger.debug(f"VehicleController: Sending motor speed {speed}")
            self.serial_handler.sendMotorSpeedCommand(speed)
            self._last_motor_speed = speed

    def _send_servo_command(self, angle: int):
        """Send servo angle command via serial handler."""
        # Clamp angle to typical servo range if not already
        clamped_angle = max(0, min(180, angle))
        if clamped_angle != self._last_servo_angle:
            logger.debug(f"VehicleController: Sending servo angle {clamped_angle}")
            self.serial_handler.sendServoAngleCommand(clamped_angle)
            self._last_servo_angle = clamped_angle

    # --- Parking Sequence Logic ---

    def start_parking_sequence(self):
        """Initiate the parallel parking sequence."""
        if self.parking_active:
            logger.warning("VehicleController: Parking already active. Cannot start new sequence.")
            return

        logger.info("VehicleController: Starting parallel parking sequence.")
        self.parking_active = True
        self.parking_state = ParkingState.APPROACHING_SPOT
        self.parking_step_counter = 0
        self.parking_start_time = time.time()
        # Reset any previous parking state variables if needed

    def _execute_parking_step(self):
        """Execute the current step of the parking sequence."""
        if not self.parking_active:
            return

        self.parking_step_counter += 1
        current_time = time.time()
        elapsed_time = current_time - self.parking_start_time

        # --- Simplified Parking State Machine ---
        # This is a basic placeholder, mirroring the C++ logic's structure.
        # A real implementation would use encoders, IMU, or precise timing/distance.
        try:
            if self.parking_state == ParkingState.APPROACHING_SPOT:
                # 1. Approach the spot slowly, aligning the rear axle
                self._send_motor_command(self._map_normalized_speed(0.3)) # Slow forward
                self._send_servo_command(90) # Straight
                logger.debug(f"Parking Step {self.parking_step_counter}: APPROACHING_SPOT")
                # Placeholder condition to move to next state
                if self.parking_step_counter > 50: # Or based on time/distance/IMU
                    logger.info("Parking: Transitioning to REVERSING_INTO_SPOT")
                    self.parking_state = ParkingState.REVERSING_INTO_SPOT
                    self.parking_step_counter = 0

            elif self.parking_state == ParkingState.REVERSING_INTO_SPOT:
                # 2. Reverse while turning steering wheel
                self._send_motor_command(self._map_normalized_speed(-0.4)) # Slow reverse
                self._send_servo_command(45) # Turn wheels (e.g., full left)
                logger.debug(f"Parking Step {self.parking_step_counter}: REVERSING_INTO_SPOT")
                # Placeholder condition
                if self.parking_step_counter > 70:
                    logger.info("Parking: Transitioning to ADJUSTING_POSITION")
                    self.parking_state = ParkingState.ADJUSTING_POSITION
                    self.parking_step_counter = 0

            elif self.parking_state == ParkingState.ADJUSTING_POSITION:
                # 3. Adjust position (e.g., straighten wheels, move forward/back)
                self._send_motor_command(self._map_normalized_speed(0.3)) # Slow forward
                self._send_servo_command(135) # Turn wheels (e.g., full right)
                logger.debug(f"Parking Step {self.parking_step_counter}: ADJUSTING_POSITION")
                # Placeholder condition
                if self.parking_step_counter > 30:
                    logger.info("Parking: Transitioning to PARKING_COMPLETE")
                    self.parking_state = ParkingState.PARKING_COMPLETE
                    self.parking_step_counter = 0

            elif self.parking_state == ParkingState.PARKING_COMPLETE:
                # 4. Finalize parking
                self._send_motor_command(0) # Stop
                self._send_servo_command(90) # Center
                logger.info("VehicleController: PARKING SEQUENCE COMPLETED.")
                self.parking_active = False # End the sequence
                # Potentially update MissionState phase to COMPLETED here or in main loop

            elif self.parking_state == ParkingState.PARKING_FAILED:
                # Handle failure (e.g., timeout, obstacle detected)
                logger.error("VehicleController: PARKING SEQUENCE FAILED.")
                self._send_motor_command(0)
                self._send_servo_command(90)
                self.parking_active = False

            # Optional: Add a timeout mechanism
            if elapsed_time > 30.0: # e.g., 30 seconds max
                logger.error("VehicleController: Parking timeout. Aborting.")
                self.parking_state = ParkingState.PARKING_FAILED
                self._execute_parking_step() # Re-enter to handle FAILED state

        except Exception as e:
            logger.error(f"VehicleController: Error during parking step execution: {e}")
            self.parking_state = ParkingState.PARKING_FAILED
            self._execute_parking_step() # Attempt to handle error state

    def is_maneuver_active(self) -> bool:
        """Check if a complex maneuver (like parking) is currently active."""
        return self.parking_active # Could extend to other maneuvers

    def emergency_stop(self):
        """Immediately send stop commands and halt maneuvers."""
        logger.critical("VehicleController: EMERGENCY STOP INITIATED!")
        self.parking_active = False # Cancel any active maneuver
        self._send_motor_command(0)
        self._send_servo_command(90)
        self.serial_handler.sendEmergencyStop() # Send dedicated E-Stop command


# Example usage (if run as script - needs mocks)
if __name__ == "__main__":
    # This would normally be integrated into the main application flow
    print("VehicleController class defined.")
